### PR TITLE
[CI] Introduce testing for IDF 5 (and other arbitrary framework versions)

### PR DESCRIPTION
**Pull request by kbx81**
# What does this implement/fix?

I tried to build this in a way that minimizes changes required as we bump the default framework versions -- all that should be required as we do this is some minor changes to files in the `tests/test_build_components` directory.

For any component that has an IDF test (for example, `test.esp32-idf.yaml`), the test script will now look for base test files with a version suffix and test with those, as well. For example, `build_components_base.esp32-idf.yaml` will continue to be the test file which uses ESPHome's default IDF framework version, but an additional file `build_components_base.esp32-idf-50.yaml` can contain a base configuration which uses an alternate IDF framework version (in this case, IDF v5.0.2 and associated PlatformIO version).

In addition, I removed the logic that would "assume" Arduino for test files with no framework in their name (hence all the file renames in this PR); this made implementing the above logic cleaner and easier and I think being explicit is generally better, anyway.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by kbx81** [comment]
Script used to do file renaming:

```python
#!/usr/bin/env python3
import os

dry_run = False
path = "tests/components"

file_renames = {
    "test.bk72xx.yaml": "test.bk72xx-ard.yaml",
    "test.esp32-c3.yaml": "test.esp32-c3-ard.yaml",
    "test.esp32-s2.yaml": "test.esp32-s2-ard.yaml",
    "test.esp32-s3.yaml": "test.esp32-s3-ard.yaml",
    "test.esp32.yaml": "test.esp32-ard.yaml",
    "test.esp8266.yaml": "test.esp8266-ard.yaml",
    "test.rp2040.yaml": "test.rp2040-ard.yaml",
}

for some_file in os.listdir(path):
    some_test_file_dir = path + "/" + some_file
    if os.path.isdir(some_test_file_dir):
        print(f"Directory: {some_test_file_dir}")
        for some_test_file in os.listdir(some_test_file_dir):
            if some_test_file in file_renames:
                new_file_name = file_renames[some_test_file]
                print(f"    {some_test_file} --> {new_file_name}")
                some_test_file = some_test_file_dir + "/" + some_test_file
                new_file_name = some_test_file_dir + "/" + new_file_name
                if not dry_run:
                    os.rename(some_test_file, new_file_name)
```

**Comment by kbx81** [comment]
At this point, I believe the only outstanding changes needed are:

- [x] https://github.com/droscy/esp_wireguard/pull/5
- [ ] https://github.com/esphome/esphome/pull/6809

File: tests/test_build_components/build_components_base.esp32-s2-idf.yaml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 esphome:
-  name: componenttestesp32s2ard
+  name: componenttestesp32s2idf
```
Comment (jesserockz) [inline tests/test_build_components/build_components_base.esp32-s2-idf.yaml:2]: :see_no_evil:

**Review by jesserockz — APPROVED** [review]
Lets hope this is not going to blanket increase the CI time from 6 hours to 12 hours if anyone touches `esphome/components/sensor/*`

**Comment by kbx81** [comment]
Ok, _no one touch_ `esphome/components/sensor/*` ...for now... 😅

**Pull request merged**